Link tree

Reference paper: 
-  Other metazoans with similar approach (Sebé-Pedrós et al., 2018): 
    - https://doi.org/10.1038/s41559-018-0575-6 
    - Code: https://github.com/sebepedroslab/Xenoturbella_sc_atlas/
    
- Trichoplax cell atlas (Najle et al., 2023): 
    - https://doi.org/10.1016/j.cell.2023.08.027
    - cell atlas website: https://sebelab.crg.eu/placozoa_cell_atlas/ 
        - https://github.com/sebepedroslab/shinyscdb
    - GitHub page: https://github.com/xgrau/placozoa-cell-type-evolution-code

- Trichoplax H2 endosymbionts, Gruber-Vodicka et al. (2019) 
    - https://doi.org/10.1038/s41564-019-0475-9

- Cell tpyes and body plan of Trichplax adhaerens, Smith et al. (2014)
    - https://doi.org/10.1016/j.cub.2014.05.046

- 
    

Bioinformatics tool:
- sratoolkit: https://hpc.nih.gov/apps/sratoolkit.html
    - https://github.com/ncbi/sra-tools
    - https://anaconda.org/bioconda/sra-tools
- cellranger:https://www.10xgenomics.com/support/software/cell-ranger/latest/analysis
- SEURAT: https://satijalab.org/seurat/
    - https://github.com/satijalab/seurat
- metacell: https://tanaylab.github.io/metacell/
    - https://github.com/tanaylab/metacell
    - https://pypi.org/project/metacells/0.8.0/
    - https://github.com/tanaylab/metacells
- mc2d: 
    - https://github.com/cran/mc2d


General helps
- scRNA-seq analysis: https://github.com/hbctraining/scRNA-seq
    - archived: https://hbctraining.github.io/scRNA-seq/lessons/04_SC_quality_control.html
- 