{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Metacell run"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "outputs": [],
   "source": [
    "# Load the MetaCell data (host cell atlas)\n",
    "load(\"E:/scTriH2/metacell/mc.scdr_TrH2_it4.Rda\")\n",
    "\n",
    "# Check what objects were loaded\n",
    "ls()\n",
    "\n",
    "# Assuming the main metacell object is called 'metacell_host', inspect it\n",
    "str(metacell_host)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, download the database from the github repository of xgrau:\n",
    "- https://github.com/xgrau/placozoa-cell-type-evolution-code\n",
    "- the directory is scatlas_results/data/scdb/\n",
    "Then, load the metacell data as in read.me of the repo\n",
    "- my data key is scdr_TrH2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "outputs": [],
   "source": [
    "library(metacell)\n",
    "setwd(\"/mnt/d/scTriH2/metacell\")\n",
    "source(\"xboc_functions.R\")\n",
    "# initialise metacell database using its relative path:\n",
    "metacell::scdb_init(\"/mnt/d/scTriH2/metacell/mc_unzipped/\", force_reinit = TRUE)\n",
    "\n",
    "# read the mat object using its ID to refer to it; there ara analogous files for metacells, etc\n",
    "mat_no_bact <- scdb_mat(\"scdr_TrH2_it2\")\n",
    "mat_drug <- scdb_mat(\"drugs_TrH2_it1\")\n",
    "\n",
    "\n",
    "# sparse matrix available in the mat@mat slot, in this case it contains x genes and y cells\n",
    "dim(mat_no_bact@mat)\n",
    "    # 13948 genes x 15788 cells\n",
    "dim(mat_drug@mat)\n",
    "sum(mat_drug@mat)\n",
    "    # 18334556 RNAs\n",
    "\n",
    "# also get the RNA counts in other libraries\n",
    "library_tag <- c(\"H2_1_ACME_10x_10kc_\", \"H2_2_ACME_10x_10kc_\", \"H2_3_ACME_10x_10kc_\", \"H2H23_4_ACME_10x_10kc_\", \"Plac02_H2_H23_10XscRNAseq_10kc\", \"H2drugs_10XscRNAseq_10kc_\" )\n",
    "library_cells_H2_1 <- grep(paste0(\"^\", \"H2drugs\"), colnames(mat_no_bact@mat), value = TRUE)\n",
    "library_mat_H2_1 <- mat_no_bact@mat[, library_cells_H2_1]\n",
    "library_total_counts <- sum(rowSums(as.matrix(library_mat_H2_1)))\n",
    "print(library_total_counts)\n",
    "    # H2_1_ACME_10x_10kc_: 3540154\n",
    "# repeat it for every barcodes\n",
    "    # H2_2_ACME_10x_10kc :  1436845\n",
    "    # H2_3_ACME_10x_10kc : 1234842\n",
    "    # H2H23_4_ACME_10x_10kc : 6440214\n",
    "    # Plac02_H2_H23_10XscRNAseq_10kc : 10977515\n",
    "    # H2drugs_10XscRNAseq_10kc : 0\n",
    "\n",
    "# Relation of baterial to host RNA counts\n",
    "library_tag <- c(\"H2_1_ACME_10x_10kc_\", \"H2_2_ACME_10x_10kc_\", \"H2_3_ACME_10x_10kc_\", \"H2H23_4_ACME_10x_10kc_\", \"Plac02_H2_H23_10XscRNAseq_10kc\", \"H2drugs_10XscRNAseq_10kc_\" )\n",
    "no_bact_count <- c(3540154, 1436845, 1234842, 6440214, 10977515, 18334556)\n",
    "bact_count <- c(28292, 5685, 4813, 18306, 15997, 29498)\n",
    "bact_count_perc <- bact_count/no_bact_count\n",
    "\n",
    "abundance_tab <- data.frame(\n",
    "  library_tag = library_tag,\n",
    "  no_bact_count = no_bact_count,\n",
    "  bact_count = bact_count,\n",
    "  bact_count_perc = bact_count_perc\n",
    ")\n",
    "write.csv(abundance_tab, \"bact_abundance.csv\", row.names = FALSE)\n",
    "\n",
    "# plot a bar plot\n",
    "library(ggplot2)\n",
    "bact_abundance <- read.csv(\"bact_abundance.csv\")\n",
    "rel_abundance_plot <- ggplot(bact_abundance, aes(x = library_tag, y = bact_count_perc)) +\n",
    "  geom_bar(stat = \"identity\", fill = \"red\", width= ) +\n",
    "  geom_text(aes(label=bact_count_perc), vjust=-0.3, size=3.5)+\n",
    "  labs(title = \"Bacterial Count Percentage by Library\",\n",
    "       x = \"Library\",\n",
    "       y = \"relative abundance of bacterial signal to total\")+\n",
    "    theme_minimal()+\n",
    "    scale_fill_brewer(palette=\"Dark2\")\n",
    "ggsave(\"bact_count_perc_barplot.png\", plot = plot, width = 10, height = 6, dpi = 300)\n",
    "\n",
    "# likewise, you may want to load a mc object containing cell-to-metacell assignments\n",
    "mc = metacell::scdb_mc(\"scdr_TrH2_it4\")\n",
    "mc_drug = metacell::scdb_mc(\"drugs_TrH2_it1\")\n",
    "# the mc@mc slot is a vector with all cells and their associated metacell\n",
    "length(mc@mc)\n",
    "# 15704 cells associated with metacells\n",
    "length(mc_drug@mc)\n",
    "\n",
    "ctt = read.table(\"annotation_mc.TrH2.it4.reordered.tsv\", header = TRUE)\n",
    "ctt_col = scr_load_cell_type_table(\"annotation_mc.TrH2.it4.reordered.tsv\", mc)\n",
    "head(ctt)\n",
    "nrow(ctt[\"metacell\"])\n",
    "# 217 metacells present\n",
    "unique(ctt[\"cell_type\"])\n",
    "                              cell_type\n",
    "1                              lipophil\n",
    "37               trans_lipophil_1_fibre\n",
    "38               trans_lipophil_1_gland\n",
    "40           trans_lipophil_1_epithelia\n",
    "41                            unknown_1\n",
    "46                          unknown_2.1\n",
    "47                          unknown_2.2\n",
    "51                          unknown_2.3\n",
    "52                          unknown_2.4\n",
    "59                                fibre\n",
    "91                    trans_fibre_gland\n",
    "92                trans_fibre_epithelia\n",
    "94                              gland_1\n",
    "120                epithelia_gland_like\n",
    "130                 epithelia_unknown_1\n",
    "131                   epithelia_ventral\n",
    "157 trans_epithelia_ventral_peptidergic\n",
    "159      trans_epithelia_dorsal_ventral\n",
    "164               epithelia_dorsal_like\n",
    "178                    epithelia_dorsal\n",
    "194             peptidergic_progenitors\n",
    "195                   peptidergic_alpha\n",
    "197                    peptidergic_beta\n",
    "198                   peptidergic_gamma\n",
    "199              peptidergic_gamma_like\n",
    "200                   peptidergic_delta\n",
    "201                 peptidergic_epsilon\n",
    "202                    peptidergic_zeta\n",
    "203                     peptidergic_eta\n",
    "205                   peptidergic_theta\n",
    "206                    peptidergic_iota\n",
    "209               peptidergic_iota_like\n",
    "211                   peptidergic_kappa\n",
    "213                  peptidergic_lambda\n",
    "214                      peptidergic_mu\n",
    "215                 peptidergic_mu_like\n",
    "216                      peptidergic_nu\n",
    "217                 peptidergic_omicron\n",
    "\n",
    "\n",
    "# find the library attatched to the barcodes: \n",
    "bc_name <- names(mc@mc)\n",
    "unique_library_names <- unique(library_names)\n",
    "num_unique_libraries <- length(unique_library_names)\n",
    "print(unique_library_names)\n",
    "# \"H2_1_ACME_10x_10kc\"             \"H2_2_ACME_10x_10kc\"\n",
    "# \"H2_3_ACME_10x_10kc\"             \"H2H23_4_ACME_10x_10kc\"\n",
    "# \"Plac02_H2_H23_10XscRNAseq_10kc\"\n",
    "library_counts <- table(library_names)\n",
    "print(library_counts)\n",
    "# do the same for the notch drug library\n",
    "# \"H2drugs_10XscRNAseq_10kc\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Secondly, the single cell data matrix (in this case in seurat object) is loaded. \n",
    "- The scRNA data are merged and processed (normalization , PCA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "outputs": [],
   "source": [
    "# Here the sc library stored as SeuratObject format loaded and analyzed\n",
    "library(Seurat)\n",
    "library(SingleCellExperiment)\n",
    "library(Matrix)\n",
    "\n",
    "parent_directory <- \"/mnt/d/scTriH2/SEURAT/extracted/1.cellRangerCount\"\n",
    "SRR_list_full <- list.dirs(parent_directory, full.names = FALSE, recursive = FALSE)\n",
    "SRR_list_full <- SRR_list_full[grepl(\"SRR*\", SRR_list_full)]\n",
    "SRR_list <- SRR_list_full[!SRR_list_full %in% c(\"SRR24886387\", \"SRR24886410\")] #, \"SRR24886416\"\n",
    "grellia_list <- list()\n",
    "tot_read_list <- list()\n",
    "gr_tot_list <- list()\n",
    "\n",
    "# to obtain relative abundance of grellia in each sample, optionally depending per metacells   \n",
    "# the total read numbers are \n",
    "\n",
    "\n",
    "for (i in 1:7) {\n",
    "    dir_path <- paste0(\"/mnt/d/scTriH2/SEURAT/extracted/1.cellRangerCount/Seurat_outputs/\", SRR_list[i],\"_output\")\n",
    "    object_name <- paste0(SRR_list[i], \"_filtered.rds\")\n",
    "    object_path <- file.path(dir_path, object_name) \n",
    "\n",
    "    # import SeuratObject to mat as bact_mat\n",
    "    seurat_object <- readRDS(object_path)\n",
    "    sce = as.SingleCellExperiment(seurat_object, assay = \"RNA\")\n",
    "    bact_mat = scm_import_sce_to_mat(sce)\n",
    "\n",
    "    ## This step is now done in seurat loop\n",
    "    # change barcode names like in names(mc@mc)\n",
    "    # colnames(bact_mat@mat) <- paste0(\"H2_1_ACME_10x_10kc_\", colnames(bact_mat@mat))\n",
    "\n",
    "    # Add up every mapped bacterial read contained in a cell (barcode)\n",
    "    grellia <- colSums(as.matrix(bact_mat@mat))\n",
    "    grellia_name <- paste0(\"grellia_\", i)\n",
    "\n",
    "    # Directly assign grellia to the list â€” no need for append()\n",
    "    grellia_list[[grellia_name]] <- grellia\n",
    "\n",
    "    # Handle missing barcodes and count bacterial signal per metacell\n",
    "    if (SRR_list[i] == \"SRR24886416\" | SRR_list[i] == \"SRR24886417\") {\n",
    "        missing_barcodes <- setdiff(names(mc_drug@mc), names(grellia))\n",
    "        grellia[missing_barcodes] <- 0   \n",
    "        grellia_total <- tapply(grellia[names(mc_drug@mc)], mc_drug@mc, sum)\n",
    "    } else {\n",
    "        missing_barcodes <- setdiff(names(mc@mc), names(grellia))\n",
    "        grellia[missing_barcodes] <- 0\n",
    "        grellia_total <- tapply(grellia[names(mc@mc)], mc@mc, sum)\n",
    "    }\n",
    "\n",
    "    # Add grellia_total to the list\n",
    "    grellia_tot_name <- paste0(\"grellia_total_\", i)\n",
    "    gr_tot_list[[grellia_tot_name]] <- grellia_total\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "outputs": [],
   "source": [
    "library(Seurat)\n",
    "# load the symbiont sc matrix\n",
    "BiocManager::install(\"SingleCellExperiment\")\n",
    "library(SingleCellExperiment)\n",
    "install.packages(\"SeuratObject\")\n",
    "summary_object_1 <- readRDS(\"/mnt/d/scTriH2/SEURAT/extracted/1.cellRangerCount/Seurat_outputs/merged_seurat_object_1.rds\")\n",
    "sce = as.SingleCellExperiment(summary_object_1, assay=\"RNA\")\n",
    "\n",
    "# or\n",
    "merged_seurat <- readRDS(\"/mnt/d/scTriH2/SEURAT/extracted/1.cellRangerCount/Seurat_outputs/merged_seurat_object_1.rds\")\n",
    "sce = as.SingleCellExperiment(merged_seurat, assay=\"RNA\")\n",
    "\n",
    "bact_mat = scm_import_sce_to_mat(sce)\n",
    "library(Matrix)\n",
    "\n",
    "# change barcode names like in names(mc@mc)\n",
    "colnames(bact_mat@mat) <- paste0(\"H2_1_ACME_10x_10kc_\", colnames(bact_mat@mat))\n",
    "\n",
    "# add up every genes contained in a cell(barcode)\n",
    "grellia <- colSums(as.matrix(bact_mat@mat))\n",
    "\n",
    "# add missing barcodes to the grellia and set values to 0 before summing\n",
    "missing_barcodes <- setdiff(names(mc@mc), names(grellia))\n",
    "grellia[missing_barcodes] <- 0\n",
    "\n",
    "# Count bacterial signal per metacell\n",
    "grellia_total <- tapply(grellia[names(mc@mc)], mc@mc, sum)\n",
    "\n",
    "\n",
    "## First, visualize how many grellia signal is contained in each metacells\n",
    "# bacterial signal per metacell, point size = log(signals per cell)\n",
    "pdf(\"bacterial_signal_per_mc.pdf\", h=8, w=8, useDingbats=F)\n",
    "\n",
    "# Scatter plot of grellia_total\n",
    "plot(grellia_total, pch=20, cex=log(grellia_total + 1), col=as.character(ctt$color))\n",
    "\n",
    "# Add text labels\n",
    "text(grellia_total, labels=names(grellia_total), col=ifelse(pmax(grellia_total)> 6,\"black\",alpha(\"black\",0)))\n",
    "\n",
    "# Extract unique cell types with over 50 bacterial signal and corresponding colors\n",
    "filtered_cells <- names(grellia_total)[grellia_total > 50]\n",
    "filtered_ctt <- ctt[filtered_cells, , drop = FALSE]\n",
    "\n",
    "unique_cell_types <- unique(filtered_ctt$cell_type)\n",
    "unique_colors <- filtered_ctt$color[match(unique_cell_types, filtered_ctt$cell_type)]\n",
    "\n",
    "\n",
    "# Add legend with unique cell types\n",
    "legend(\"topright\", legend=unique_cell_types, col=unique_colors, pch=20, border=\"black\", cex=1)\n",
    "\n",
    "dev.off()\n",
    "\n",
    "\n",
    "# sc bacterial signal 2D projections, absolute counts\n",
    "mc2d <- scdb_mc2d(\"scdr_TrH2_it4\")\n",
    "grellia_mc2d_abs <- scp_plot_gene_2d_metacell_bacteria(mc2d,mc,grellia_total,out_fn=\"Grellia_2d_abs.png\",plot_mc=F,log=F)\n",
    "grellia_mc2d_mc_abs <- scp_plot_gene_2d_metacell_bacteria(mc2d,mc,grellia_total,out_fn=\"Grellia_2d_mc_abs.p\n",
    "ng\",plot_mc=T,log=F)\n",
    "\n",
    "# metacell visualization\n",
    "mc_col <- as.vector(ctt$color)\n",
    "mc2d_plot <- mcell_mc2d_plot(\"scdr_TrH2_it4\", legend_pos =\"bottomright\", colors = mc_col, fig_fn=\"mc2d_plot.png\")\n",
    "\n",
    "\n",
    "grellia_mc2d_mc_col_abs <- scp_plot_gene_2d_metacell_bacteria(mc2d_plot,mc,grellia_total, out_fn=\"Grellia_2d_mc_abs.png\",plot_mc=T,log=F)\n",
    "\n",
    "\n",
    "mat_no_bact_counts <- as.matrix(mat_no_bact@mat)\n",
    "host_counts <- as.matrix(mc@mc_fp)\n",
    "mc_counts <- as.matrix(mat_no_bact_counts) %*% t(as.matrix(mc_fp)) # genes x metacells matrix\n",
    "dim(mat_no_bact_counts)\n",
    "dim(host_counts)\n",
    "\n",
    "# Check if the genes are the same and in the same order\n",
    "all(rownames(mat_no_bact_counts) == rownames(host_counts))\n",
    "\n",
    "# If not, align them\n",
    "mat_no_bact_counts <- mat_no_bact_counts[rownames(host_counts), ]\n",
    "\n",
    "# Assuming 'mc@mc' is a vector of metacell assignments for each cell\n",
    "# and 'mat_no_bact_counts' is a gene-by-cell matrix\n",
    "\n",
    "# Aggregate host counts by metacell\n",
    "mc_counts <- t(rowsum(t(mat_no_bact_counts), group = mc@mc))\n",
    "\n",
    "# Check dimensions\n",
    "dim(mc_counts)  # Should be genes x metacells\n",
    "\n",
    "mat_no_bact_counts <- as.matrix(mat_no_bact_counts)\n",
    "storage.mode(mat_no_bact_counts) <- \"numeric\"\n",
    "\n",
    "mc_fp <- as.matrix(mc_fp)\n",
    "storage.mode(mc_fp) <- \"numeric\"\n",
    "\n",
    "\n",
    "mc_counts <- as.matrix(mat_no_bact_counts) %*% t(as.matrix(mc_fp)) # genes x metacells matrix\n",
    "\n",
    "\n",
    "\n",
    "mc_sizes=colSums(mc_counts)\n",
    "grellia_norm=grellia_total*100/(grellia_total+mc_sizes)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "outputs": [],
   "source": [
    "# or also "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "After counting:\n",
    "Generate figures for the metacell bacterial signal\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "outputs": [],
   "source": [
    "# metacell bacterial signal\n",
    "pdf(\"bacterial_signal_per_mc.pdf\",h=8,w=8,useDingbats=F)\n",
    "plot(grellia_norm, pch=20, cex=4, col=as.character(ctt$color))\n",
    "text(grellia_norm, labels=names(grellia_norm), col=ifelse(pmax(grellia_norm)>4,\"black\",alpha(\"black\",0)))\n",
    "dev.off()\n",
    "\n",
    "# sc bacterial signal 2D projections\n",
    "mc2d <- scdb_mc2d(mc2d_id=\"scdr_TrH2_it4\")\n",
    "grellia_mc2d <- scp_plot_gene_2d_metacell_bacteria(mc2d,mc,grellia_norm,out_fn=\"Grellia_2d.png\",plot_mc=F,log=F)\n",
    "\n",
    "# fraction of infected/bacterial-containing cells per animal\n",
    "cell_animal=as.vector(mat@cell_metadata[names(mc@mc),\"dataset\"])\n",
    "names(cell_animal)=names(mc@mc)\n",
    "frac_grellia_cells=table(cell_animal[grellia_positive_cells])*100/table(cell_animal)\n",
    "\n",
    "pdf(\"Frac_infected_cells_per_animal.pdf\",h=10,w=6,useDingbats=F)\n",
    "par(mfrow=c(2,1))\n",
    "barplot(frac_grellia_cells,ylab=\"% infected cells\",col=\"gray30\",ylim=c(0,10),main=\"Grellia\")\n",
    "dev.off()\n"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
